[Speculation] Updated SpeculationPlacement

shundroid opened the pull request:
**Summary of Changes:**

- **Buffer Placement Update**
  - Removed buffer placement logic from `SpeculationPlacement`.
  - Now relies on the general `HandshakeBufferPlacementCustom` for buffer insertion.
- **FIFO Depth Support in Speculation Units**
  - Added support for configuring FIFO depths of `Speculator` and `SaveCommit` units via `spec.json`.
  - The FIFO depth value is:
    - Stored as an attribute in `SpeculatorOp` and `SpecSaveCommitOp` (based on @schilkp's suggestion).
    - Propagated to `hw.parameters` during the `HandshakeToHW` conversion (to be moved to a dedicated MLIR interface in the future).
- **Refactoring**
  - Separated `placeUnits<T>` logic for save-commit units to allow FIFO depth customization:
    - Split into `placeCommits` and `placeSaveCommits`.
    - Commented out `placeUnits<SpecSaveOp>` and added a validation to skip save unit placement entirely.
  - Cleaned up method structure and naming:
    - Restructured `prepareAndPlaceSaveCommits` to `generateSaveCommitCtrl`.

shundroid commented on integration-test/subdiag/spec.json:
```diff
@@ -1,6 +1,8 @@
     "operation-name": "fork3",
-    "operand-idx": 0
-  }
+    "operand-idx": 0,
+    "fifo-depth": 19
+  },
+  "save-commits-fifo-depth": 19
```
Only this benchmark needs the fifo depth of 19😂

shundroid commented on experimental/lib/Transforms/Speculation/PlacementFinder.cpp:
```diff
@@ -33,14 +33,6 @@ PlacementFinder::PlacementFinder(SpeculationPlacements &placements)
   assert(specPos.getOwner() && "Speculator position is undefined");
 }
 
-void PlacementFinder::clearPlacements() {
-  // Speculator and buffer positions are manually set
-  OpOperand &specPosition = placements.getSpeculatorPlacement();
-  llvm::DenseSet<OpOperand *> bufferPositions =
-      this->placements.getPlacements<handshake::BufferOp>();
-  this->placements = SpeculationPlacements(specPosition, bufferPositions);
-}
```
`clearPlacement` used to discard all manual configurations except for `specPosition` and `bufferPositions`. The `placements` variable here is a reference from the `HandshakeSpeculation` pass, and this behavior of clearing the object owned by the pass was actually quite confusing (I got tripped up by it multiple times). It wipes unrelated manual configurations for no clear reason, so I've removed that behavior now.

murphe67 commented on experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp:
```diff
@@ -422,7 +402,28 @@ LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  // Set the FIFO depth attribute for each SaveCommit
+  specOp->getParentOp()->walk([&](Operation *op) {
+    if (auto saveCommitOp = dyn_cast<handshake::SpecSaveCommitOp>(op)) {
+      SmallVector<NamedAttribute> scHWParams;
+      scHWParams.emplace_back(
+          /*key=*/builder.getStringAttr(FIFO_DEPTH_ATTR_NAME),
+          /*value=*/builder.getIntegerAttr(
+              builder.getIntegerType(/*width=*/32, /*isSigned=*/false),
+              static_cast<int64_t>(placements.getSaveCommitsFifoDepth())));
+      saveCommitOp->setAttr(RTL_PARAMETERS_ATTR_NAME,
+                            builder.getDictionaryAttr(scHWParams));
```
does this assume hw.parameters does not exist already? 
this is probably a pretty safe assumption, but how complicated would it be to work either way?

murphe67 commented on experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp:
```diff
@@ -422,7 +402,28 @@ LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  }
+
+  // Set the FIFO depth attribute for each SaveCommit
+  specOp->getParentOp()->walk([&](Operation *op) {
+    if (auto saveCommitOp = dyn_cast<handshake::SpecSaveCommitOp>(op)) {
```
same comment as on the other PR- at a glance this feels like a weird way to get the save commits, but I am not familiar enough with MLIR to know how typical this kind of approach is

shundroid replied:
I think I'll make placeSaveCommits function and call it, instead of using template placeUnits function.

Inside it we register fifo depth immediately after the op is created.

In this way we can assure hw.parameters doesn't exist more safely.

After I implement it in this way I'll get back to you!

murphe67 reviewed (commented):
Looks great, thanks for the good work Shun!

shundroid commented:
@murphe67 I went through the previous implementation and made large updates.

Based on @schilkp's suggestion, I added a `fifoDepth` attribute to both `SpeculatorOp` and `SpecSaveCommitOp`, and I'm propagating it to `hw.parameters` during `HandshakeToHW`. This will eventually move to an MLIR interface method, but do you think I should put this logic in `RTL.cpp` for now?

I also removed the entire `funcOp` traversal.

See the updated pull request description for more details!

shundroid commented on experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp:
```diff
@@ -85,64 +85,6 @@ struct HandshakeSpeculationPass
-    inheritBB(dstOp, newOp);
-
-    // Connect the new Operation to dstOp
-    // Note: srcOpResult.replaceAllUsesExcept cannot be used here
-    // The following bug may occur in most cases:
-    // (a) Consider a scenario where a control value from a buffer is passed to
-    // the control branch.
-    // (b) Simultaneously, a speculator uses the same control value from the
-    // buffer as a trigger signal.
-    // (c) A save-commit unit is positioned on the edge from the buffer to the
-    // control branch.
-    // (d) If we apply replaceAllUsesExcept to the value referenced by the
-    // save-commit unit, the speculator will also be placed after the
-    // save-commit unit, which is undesirable.
-    operand->set(newOp.getResult());
```
I removed this long comment in the new methods, because `operand->set` is a natural choice, rather than `replaceAllUsesExcept`.

murphe67 commented on experimental/include/experimental/Transforms/Speculation/SpeculationPlacement.h:
```diff
@@ -44,16 +44,17 @@ class SpeculationPlacements {
-  SpeculationPlacements(OpOperand &speculatorPosition,
-                        llvm::DenseSet<OpOperand *> &bufferPositions)
-      : speculator(&speculatorPosition), buffers(bufferPositions){};
+  /// Initializer with operands specifying the speculator position
```
```suggestion
  /// Initializer with operand specifying the speculator position
```

murphe67 reviewed (approved):
Yes, this all seems very sensible to me now- the attribute approach is quite obviously better here, thanks for making the switch

Thanks for all the good work Shun!
